[master] Build for CentOS 9.

Romain-Geissler-1A opened the pull request:
This shall be backported to branch 20.10 too.

Romain-Geissler-1A commented:
Note: this will required #633 to be merged first.

Romain-Geissler-1A commented:
Note: when this will be merged and https://github.com/docker/containerd-packaging/pull/270 will be merged/delivered, I have the local changes to build for CentOS 9 too.

thaJeztah commented on rpm/SPECS/docker-ce.spec:
```diff
@@ -22,7 +22,10 @@ Requires: container-selinux >= 2:2.74
 Requires: libseccomp >= 2.3
 Requires: systemd
 Requires: iptables
+%if ! 0%{?rhel}
```
Should we make this conditional based on `rhel` (/centos) version? If it's not deprecated in rhel7 and centos7, (or 8) we could make it something like;

```suggestion
%if 0%{?rhel} == 7 || 0%{?rhel} == 8
```

thaJeztah replied:
Using straight (`==`) comparisons to account for other distros who would like not have the `rhel` macro (but double check if my suggestion is correct 😅)

Romain-Geissler-1A replied:
Well, libcgroup is deprecated in RHEL 7 already: https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/7/html/resource_management_guide/chap-using_libcgroup_tools

> The libcgroup package, which was the main tool for cgroup management in previous versions of Red Hat Enterprise Linux, is now deprecated. To avoid conflicts, do not use libcgroup tools for default resource controllers (listed in [Available Controllers in Red Hat Enterprise Linux 7](https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/7/html/resource_management_guide/br-Resource_Controllers_in_Linux_Kernel#itemlist-Available_Controllers_in_Red_Hat_Enterprise_Linux_7)) that are now an exclusive domain of systemd. This leaves a limited space for applying libcgroup tools, use it only when you need to manage controllers not currently supported by systemd, such as net_prio.

thaJeztah replied:
The devil can be in the detail

> This leaves a limited space for applying libcgroup tools, use it only when you need to manage controllers not currently supported by systemd, such as net_prio.

so I think we should consider keeping it (for now) for the versions that still carry the package

Romain-Geissler-1A replied:
Ok, then. I used the expression `%if 0%{?rhel} <= 8` does it fly for you ?

Romain-Geissler-1A commented:
Note that another prerequisite to building for RHEL/CentOS 9 is considering something like https://github.com/docker/containerd-packaging/pull/270 (which is failing the CI, but since external contributors can't see the Jenkins logs, I don't know it's failing).

Romain-Geissler-1A commented:
Ping

Romain-Geissler-1A commented:
Ping

Romain-Geissler-1A commented:
Ping

Romain-Geissler-1A commented:
@thaJeztah Pull request rebased and updated to adopt your style as in https://github.com/docker/containerd-packaging/pull/283. I have also modified the pull request purpose, now it removes libcgroup dependency for RHEL >= 9 and also builds a package for CentOS 9 at the same time. Most likely the CI will remain red until a containerd package for CentOS 9 is delivered in the official docker repositories.

thaJeztah commented on rpm/centos-8/Dockerfile:
```diff
@@ -18,27 +18,23 @@ ARG SUITE
 ENV DISTRO=${DISTRO}
 ENV SUITE=${SUITE}
 
-# In aarch64 (arm64) images, the altarch repo is specified as repository, but
-# failing, so replace the URL.
-RUN if [ -f /etc/yum.repos.d/CentOS-Stream-Sources.repo ]; then sed -i 's/altarch/centos/g' /etc/yum.repos.d/CentOS-Stream-Sources.repo; fi
-
```
This one was an odd one; I know we added it at some point, then we removed it (because it appeared it didn't have the `altarch` anymore); https://github.com/docker/docker-ce-packaging/pull/416#discussion_r367891915

But later, we had to add it back, because for some reason, our pipeline started failing again on this; see https://github.com/docker/docker-ce-packaging/pull/446

Could it be these can be set dynamically, depending on what mirror is hit?

thaJeztah commented on rpm/centos-8/Dockerfile:
```diff
@@ -18,27 +18,23 @@ ARG SUITE
 ENV DISTRO=${DISTRO}
 ENV SUITE=${SUITE}
 
-# In aarch64 (arm64) images, the altarch repo is specified as repository, but
-# failing, so replace the URL.
-RUN if [ -f /etc/yum.repos.d/CentOS-Stream-Sources.repo ]; then sed -i 's/altarch/centos/g' /etc/yum.repos.d/CentOS-Stream-Sources.repo; fi
-
```
This one was an odd one; I know we added it at some point, then we removed it (because it appeared it didn't have the `altarch` anymore); https://github.com/docker/docker-ce-packaging/pull/416#discussion_r367891915

But later, we had to add it back, because for some reason, our pipeline started failing again on this; see https://github.com/docker/docker-ce-packaging/pull/446

Could it be these can be set dynamically, depending on what mirror is hit?

Romain-Geissler-1A replied:
I thought this was some old hack no longer necessary. But if you say it's still needed, I will put it back.

thaJeztah replied:
The honest answer: "I don't know!" 😂 (I also thought it was redundant, and then... all of a sudden, it failed in one of our internal pipelines)

Perhaps we should keep it "just to be safe" (it looks like it doesn't hurt to have it, and would be a no-op if it's not needed).

thaJeztah commented:
Overall; changes look good to me; I'm running a build locally, as CI doesn't pick up changes in Jenkinsfiles for users without write-access

thaJeztah commented:
Working on a (temporary) fix for failing CI; looks like someone's vanity URL went down; https://github.com/docker/docker-ce-packaging/pull/684

thaJeztah commented:
I kicked CI, but looks like GitHub actions doesn't do a merge before running again (whereas jenkins does).

I tested this locally, and it looks to be working as expected

thaJeztah reviewed (approved):
LGTM, thanks!
